### PR TITLE
fix: How do I create Organization API Key with Organization Billing Admin permission and Project Read Only for projects

### PR DESCRIPTION
**Pull request by andreaangiolillo**
## Description

JIRA ticket: [INTMDB-898](https://jira.mongodb.org/browse/INTMDB-898)

Link to any related issue(s): https://github.com/mongodb/terraform-provider-mongodbatlas/issues/1278

## Type of change:

- [x] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code

## Further comments

I locally tested the provider with the following configuration files and checked that the resource was successfully created, the `terraform plan` did not return any changes and the `terraform destroy` worked as expected.

```terraform
resource "mongodbatlas_project" "atlas-project" {
  name   = "Issue-1278"
  org_id = "60ddf55c27a5a20955a707d7"
}

resource "mongodbatlas_project_api_key" "api_1" {
  description = "test api_key multi"
  project_id  = mongodbatlas_project.atlas-project.id

  project_assignment {
    project_id = mongodbatlas_project.atlas-project.id
    role_names = ["ORG_BILLING_ADMIN", "GROUP_READ_ONLY"]
  }

  project_assignment {
    project_id = "63dcfc256af00a5934e60924"
    role_names = ["GROUP_READ_ONLY"]
  }

  project_assignment {
    project_id = "64c23af6f133166c39176cbf"
    role_names = ["GROUP_OWNER"]
  }
}
```

```terraform
resource "mongodbatlas_project" "atlas-project" {
  name   = "Issue-1278"
  org_id = "60ddf55c27a5a20955a707d7"
}

resource "mongodbatlas_project_api_key" "api_1" {
  description = "test api_key multi"
  project_id  = mongodbatlas_project.atlas-project.id

  project_assignment {
    project_id = mongodbatlas_project.atlas-project.id
    role_names = ["GROUP_OWNER"]
  }

  project_assignment {
    project_id = "63dcfc256af00a5934e60924"
    role_names = ["GROUP_READ_ONLY"]
  }

  project_assignment {
    project_id = "64c23af6f133166c39176cbf"
    role_names = ["GROUP_OWNER"]
  }
}
```

```bash
Running tool: /Users/andrea.angiolillo/.asdf/shims/go test -timeout 300000s -run ^TestAccConfigRSProjectAPIKey_OrgRoles$ github.com/mongodb/terraform-provider-mongodbatlas/mongodbatlas

ok  	github.com/mongodb/terraform-provider-mongodbatlas/mongodbatlas	20.060s
```

### CODE REVIEW

File: mongodbatlas/resource_mongodbatlas_project_api_key_test.go [hunk 1]
```diff
@@ -268,3 +304,37 @@ func testAccMongoDBAtlasProjectAPIKeyConfigMultiple(orgID, projectName, descript
...
 }
+
+func testAccMongoDBAtlasProjectAPIKeyConfigOrgRoles(orgID, firstProjectName, secondProjectName, description string) string {
+	return fmt.Sprintf(`
+		resource "mongodbatlas_project" "test" {
+			name   = %[2]q
+			org_id = %[1]q
+		}
+
+		resource "mongodbatlas_project" "testProject" {
```
Comment (maastha) [inline mongodbatlas/resource_mongodbatlas_project_api_key_test.go:315]: Is this used?

File: mongodbatlas/resource_mongodbatlas_project_api_key_test.go [hunk 2]
```diff
@@ -268,3 +304,37 @@ func testAccMongoDBAtlasProjectAPIKeyConfigMultiple(orgID, projectName, descript
...
+			name   = %[3]q
+			org_id = %[1]q
+		}
+
+		resource "mongodbatlas_project_api_key" "test" {
+			project_id     = mongodbatlas_project.test.id
+			description  = %[4]q
+			project_assignment  {
```
Comment (maastha) [inline mongodbatlas/resource_mongodbatlas_project_api_key_test.go:323]: Can we please update/add tests with multiple `project_assignment`s since it is a TypeSet? None of the tests for this resource seem to do be doing that to correctly test updates/reads etc.

File: website/docs/r/project_api_key.html.markdown [hunk 3]
```diff
@@ -35,13 +35,42 @@ resource "mongodbatlas_project_api_key" "test" {
...
+resource "mongodbatlas_project" "atlas-project" {
+  name   = "ProjectTest"
+  org_id = "60ddf55c27a5a20955a707d7"
+}
+
+resource "mongodbatlas_project_api_key" "api_1" {
+  description = "test api_key multi"
+  project_id  = mongodbatlas_project.atlas-project.id
+
+  project_assignment {
```
Comment (Zuhairahmed) [inline website/docs/r/project_api_key.html.markdown:58]:
```suggestion
  // NOTE: The `project_id` of the first `project_assignment` element must be the same as the `project_id` of the resource.
  project_assignment {
```

File: website/docs/r/project_api_key.html.markdown [hunk 4]
```diff
@@ -35,13 +35,42 @@ resource "mongodbatlas_project_api_key" "test" {
...
+    project_id = mongodbatlas_project.atlas-project.id
+    role_names = ["ORG_READ_ONLY", "ORG_BILLING_ADMIN", "GROUP_READ_ONLY"]
```
Comment (Zuhairahmed) [inline website/docs/r/project_api_key.html.markdown]: Can you clarify if order of elements in `role_names` array matters? if so what does order represent?

**Review by Zuhairahmed — COMMENTED** [review]
Few doc comments below, also @andreaangiolillo can you confirm that this update includes no breaking changes?

File: website/docs/r/project_api_key.html.markdown (hunk 4 above)
Comment (andreaangiolillo) [reply website/docs/r/project_api_key.html.markdown]: no, the order does not matter

**Review by Zuhairahmed — APPROVED** [review]

File: mongodbatlas/resource_mongodbatlas_project_api_key_test.go (hunk 1 above)
Comment (andreaangiolillo) [reply mongodbatlas/resource_mongodbatlas_project_api_key_test.go:315]: lol no I forgot to add the other `project_assignments` 🤦

Comment (andreaangiolillo) [reply mongodbatlas/resource_mongodbatlas_project_api_key_test.go:315]: updated

File: mongodbatlas/resource_mongodbatlas_project_api_key_test.go (hunk 2 above)
Comment (andreaangiolillo) [reply mongodbatlas/resource_mongodbatlas_project_api_key_test.go:323]: updated

**Comment by andreaangiolillo** [comment]
> can you confirm that this update includes no breaking changes?

@Zuhairahmed I tested the changes with the following configuration which contains the examples in the [doc](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/project_api_key) and they were all successful. Is there any other scenario that you think I am missing? Thanks

```terraform
resource "mongodbatlas_project" "atlas-project" {
  name   = "Issue-1278"
  org_id = "60ddf55c27a5a20955a707d7"
}

resource "mongodbatlas_project_api_key" "test1" {
  description = "Description of your API key"
  project_id  = mongodbatlas_project.atlas-project.id
  role_names  = ["GROUP_OWNER"]
}

resource "mongodbatlas_project_api_key" "test22" {
  description = "Description of your API key"
  project_id  = mongodbatlas_project.atlas-project.id

  project_assignment {
    project_id = mongodbatlas_project.atlas-project.id
    role_names = ["GROUP_READ_ONLY", "GROUP_OWNER"]
  }

  project_assignment {
    project_id = "63dcfc256af00a5934e60924"
    role_names = ["GROUP_READ_ONLY"]
  }
}

resource "mongodbatlas_project_api_key" "api_1" {
  description = "test api_key multi"
  project_id  = mongodbatlas_project.atlas-project.id

  project_assignment {
    project_id = mongodbatlas_project.atlas-project.id
    role_names = ["ORG_BILLING_ADMIN", "GROUP_READ_ONLY"]
  }

  project_assignment {
    project_id = "63dcfc256af00a5934e60924"
    role_names = ["GROUP_READ_ONLY"]
  }

  project_assignment {
    project_id = "64c23af6f133166c39176cbf"
    role_names = ["GROUP_OWNER"]
  }
}
```

**Review by maastha — APPROVED** [review]
LGTM

File: mongodbatlas/resource_mongodbatlas_project_api_key.go [hunk 5]
```diff
@@ -13,6 +13,14 @@ import (
 	matlas "go.mongodb.org/atlas/mongodbatlas"
 )
 
+const (
+	orgRolePrefix     = "ORG_"
+	projectRolePrefix = "GROUP_"
+	orgReadOnlyRole   = "ORG_READ_ONLY"
+)
+
+var orgRoleProvided = false
```
Comment (AgustinBettati) [inline mongodbatlas/resource_mongodbatlas_project_api_key.go]: is there a risk in having this global variable being used in separate operations? Set during create, and then used during read and delete.

File: website/docs/r/project_api_key.html.markdown [hunk 6]
```diff
@@ -56,6 +86,7 @@ List of Project roles that the Programmatic API key needs to have. `project_assi
 * `project_id` - (Required) Project ID to assign to Access Key
 * `role_names` - (Required) List of Project roles that the Programmatic API key needs to have. Ensure you provide: at least one role and ensure all roles are valid for the Project. You must specify an array even if you are only associating a single role with the Programmatic API key. The [MongoDB Documentation](https://www.mongodb.com/docs/atlas/reference/user-roles/#project-roles) describes the valid roles that can be assigned.
 
+~> **NOTE:** The `project_id` of the first `project_assignment` element must be the same as the `project_id` of the resource.
```
Comment (AgustinBettati) [inline website/docs/r/project_api_key.html.markdown:89]: Is it worth mentioning that organization level roles of the api key can only be defined in the first project assignment?

Comment (andreaangiolillo) [reply website/docs/r/project_api_key.html.markdown:89]: yes, it makes sense. Thanks

Updated

File: mongodbatlas/resource_mongodbatlas_project_api_key.go (hunk 5 above)
Comment (andreaangiolillo) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go]: no, I don't this is an issue but I renamed the var to make it more specific to the project API key resource

Comment (andreaangiolillo) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go]: updated

Comment (AgustinBettati) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go]: My only doubt is that this might work during acceptance tests as the test framework maintains one plugin gRPC server for the duration of each test case, while in normal Terraform operations the plugin server starts and stops ([reference](https://developer.hashicorp.com/terraform/plugin/sdkv2/testing/acceptance-tests#how-acceptance-tests-work)).
Approving the PR, I would just try a local build to be sure org roles are supported with no issue.

**Review by AgustinBettati — APPROVED** [review]

File: mongodbatlas/resource_mongodbatlas_project_api_key.go (hunk 5 above)
Comment (andreaangiolillo) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go]: > is there a risk in having this global variable being used in separate operations? Set during create, and then used during read and delete.

I think I misread this comment the first time. The reason why I defined this as a global var was to use it in the create and read operations as I need to know if the user has provided any ORG roles in the config:

The main reason for this is that the CREATE endpoint creates an org API key with `ORG_READ_ONLY` role **if the user did not provide any ORG_ROLES as input**. In this scenario, the READ will return the `ORG_READ_ONLY` role (in addition to the `GROUP_*` roles) even if the user provided only GROUP roles 


> while in normal Terraform operations the plugin server starts and stops ([reference](https://developer.hashicorp.com/terraform/plugin/sdkv2/testing/acceptance-tests#how-acceptance-tests-work)).

In this case, the READ operation is called as last step of the CREATE so we do not hit this issue

Comment (AgustinBettati) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go]: I understand the use case. My concern is if there is risk in this global state being lost (simply reset to false) during different operations. Quoting from the referenced link:

> In theory, it is possible for providers to carry internal state between operations **during tests** - but providers would have to go out of their way (and the SDK's public API) to do this.

My approach would be to just try a local build and define an `ORG_READ_ONLY` role. If we have no issues during apply and following plan operations we should be good.

Comment (andreaangiolillo) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go]: I tested the changes with the Terraform configurations provided in the description of the PR ( `ORG_READ_ONLY` is there) and in this comment https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1369#issuecomment-1671654985. Let me know if I should test locally another scenario

Comment (AgustinBettati) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go]: I tested with the following example:
```
resource "mongodbatlas_project_api_key" "api_1" {
  description = "test api_key multii"
  project_id  = "<project_id>"

  project_assignment {
    project_id = "<project_id>"
    role_names = ["ORG_READ_ONLY", "GROUP_READ_ONLY"]
  }
}
```

- `terraform apply` creates the api key successfully.
- then running `terraform plan` outputs incorrect diff showing that "ORG_READ_ONLY" will be added.

I believe the `projectAPIKeyOrgRoleProvided` is not preserving its value during the create and following read operation done in `terraform plan`, so it filters out the org role.

Let me know if you are able to reproduce the same, it is indeed an edge case but might be worth addressing.

Comment (andreaangiolillo) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go]: I am not able to reproduce the issue

apply
```terraform
tfa                                                                                                  6s  1.20.3  3.1.1  16.18.0 bazel 5.4.0 ﴃ cloud-local
mongodbatlas_project.atlas-project: Refreshing state... [id=64d4acf6bf469a0c1549c869]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # mongodbatlas_project_api_key.api_1 will be created
  + resource "mongodbatlas_project_api_key" "api_1" {
      + api_key_id  = (known after apply)
      + description = "test api_key multii"
      + id          = (known after apply)
      + private_key = (sensitive value)
      + project_id  = "64d4acf6bf469a0c1549c869"
      + public_key  = (known after apply)

      + project_assignment {
          + project_id = "64d4acf6bf469a0c1549c869"
          + role_names = [
              + "GROUP_READ_ONLY",
              + "ORG_READ_ONLY",
            ]
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.
mongodbatlas_project_api_key.api_1: Creating...
mongodbatlas_project_api_key.api_1: Creation complete after 1s [id=YXBpX2tleV9pZA==:NjRkNGFkMWJlODA3YjAzNWZjMjlmNTY2-cHJvamVjdF9pZA==:NjRkNGFjZjZiZjQ2OWEwYzE1NDljODY5]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```


plan:
```terraform
tf plan                                                                                              3s  1.20.3  3.1.1  16.18.0 bazel 5.4.0 ﴃ cloud-local
mongodbatlas_project.atlas-project: Refreshing state... [id=64d4acf6bf469a0c1549c869]
mongodbatlas_project_api_key.api_1: Refreshing state... [id=YXBpX2tleV9pZA==:NjRkNGFkMWJlODA3YjAzNWZjMjlmNTY2-cHJvamVjdF9pZA==:NjRkNGFjZjZiZjQ2OWEwYzE1NDljODY5]

No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed
```

Comment (andreaangiolillo) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go]: I was able to reproduce the issue: the reason why it did not "fail" during the plan is that I was using the provider in debug mode. Thanks for the discussion here. I will see if I find a way around

Comment (andreaangiolillo) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go]: Unfortunately, I have not found a way to allow the use of `ORG_READ_ONLY` since it is automatically added when ORG roles are not provided. As a result, I updated the logic to now allow the `ORG_READ_ONLY`  for project assignments and update the documentation to explain why. This should not be a breaking change since customers were not able to use this resource with org roles.

File: website/docs/r/project_api_key.html.markdown [hunk 7]
```diff
@@ -56,6 +86,12 @@ List of Project roles that the Programmatic API key needs to have. `project_assi
...
+
+~> **NOTE:** The organization level roles can be defined only in the first `project_assignment` element.
+
+~> **NOTE:** The `ORG_READ_ONLY` role at the organization level is invalid in this context. When the `project_assignment``` lacks organizational roles, the `mongodbatlas_project_api_key` resource generates an organization API key with the `ORG_READ_ONLY` role and associates it with `GROUP_*` roles. Consequently, the resource does not permit the use of `ORG_READ_ONLY` to ensure consistency between configuration and state.
```
Comment (andreaangiolillo) [inline website/docs/r/project_api_key.html.markdown:93]: FYI

File: mongodbatlas/resource_mongodbatlas_project_api_key.go [hunk 8]
```diff
@@ -496,3 +521,17 @@ func getAPIProjectAssignments(ctx context.Context, conn *matlas.Client, apiKeyOr
...
+	// the role ORG_READ_ONLY. We want to remove this from the state
+	// to avoid differences between config and state
+	for _, r := range apiKey.RoleNames {
+		if r == orgReadOnlyRole {
+			return fmt.Errorf(`%[1]s is not an allowed role for the resource. Remove %[1]s from the roles and run terraform apply again. Check out the resource documentation to know more`, orgReadOnlyRole)
+		}
+	}
+
+	return nil
+}
```
Comment (andreaangiolillo) [inline mongodbatlas/resource_mongodbatlas_project_api_key.go:537]: validating that the user did not provide `ORG_READ_ONLY`

**Review by AgustinBettati — APPROVED** [review]
LGTM

**Review by Zuhairahmed — APPROVED** [review]
